pkg/cloudprovider/provider/vultr/provider.go

```diff
@@ -0,0 +1,374 @@
+/*
+Copyright 2019 The Machine Controller Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vultr
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	cloudprovidererrors "github.com/kubermatic/machine-controller/pkg/cloudprovider/errors"
+
+	"github.com/kubermatic/machine-controller/pkg/apis/cluster/common"
+	clusterv1alpha1 "github.com/kubermatic/machine-controller/pkg/apis/cluster/v1alpha1"
+	"github.com/kubermatic/machine-controller/pkg/cloudprovider/instance"
+	vultrtypes "github.com/kubermatic/machine-controller/pkg/cloudprovider/provider/vultr/types"
+	cloudprovidertypes "github.com/kubermatic/machine-controller/pkg/cloudprovider/types"
+	"github.com/kubermatic/machine-controller/pkg/providerconfig"
+	providerconfigtypes "github.com/kubermatic/machine-controller/pkg/providerconfig/types"
+	"github.com/vultr/govultr/v2"
+	"golang.org/x/oauth2"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type provider struct {
+	configVarResolver *providerconfig.ConfigVarResolver
+}
+
+// New returns a new vultr provider.
+func New(configVarResolver *providerconfig.ConfigVarResolver) cloudprovidertypes.Provider {
+	return &provider{configVarResolver: configVarResolver}
+}
+
+type Config struct {
+	APIKey string
+	Region string
+	Plan   string
+	OsID   string
+	Tags   []string
+}
+
+func getIDForOS(os providerconfigtypes.OperatingSystem) (int, error) {
+	switch os {
+	case providerconfigtypes.OperatingSystemUbuntu:
+		return 1743, nil
+		// name: CentOS 7 x64
+	case providerconfigtypes.OperatingSystemCentOS:
+		return 167, nil
+		// name: Rocky Linux 9 x64
+	case providerconfigtypes.OperatingSystemRockyLinux:
+		return 1869, nil
+	}
+	return 0, providerconfigtypes.ErrOSNotSupported
+}
+
+func getClient(apiKey string) *govultr.Client {
+	config := &oauth2.Config{}
+	ctx := context.Background()
```

Comment:
Please propagate context from the parent methods instead of using `context.Background()`.